Documentation/Porting.md

```diff
@@ -0,0 +1,249 @@
+ static void enumerateTypeMetadataSections(const SectionEnumerator& body) {}
+ #endif
+```
+
+## C++ stub implementations
+
+Some symbols defined in C and C++ headers, especially "complex" macros, cannot
+be represented in Swift. The `_TestingInternals` module includes a header file,
+`Stubs.h`, where you can define thin wrappers around these symbols that are
+visible to Swift. For example, to use timers on Classic, you'll need to call
+`NewTimerUPP()` to define the timer's callback, but that symbol is sometimes
+declared as a macro and cannot be called from Swift. You can add a stub function
+to `Stubs.h`:
+
+```diff
+--- a/Sources/_TestingInternals/include/Stubs.h
++++ b/Sources/_TestingInternals/include/Stubs.h
+
++#if defined(macintosh)
++static TimerUPP swt_NewTimerUPP(TimerProcPtr userRoutine) {
++  return NewTimerUPP(userRoutine);
++}
++#endif
+```
+
+Stub functions should generally be `static` to allow for inlining and when
+possible should be named to match the symbols they wrap.
+
+## Unavailable features
+
+You may find that some feature of C++, Swift, or Swift Testing cannot be ported
+to your target platform. For example, Swift Testing's `FileHandle` type includes
+an `isTTY` property to determine if a file handle refers to a pseudoterminal,
+but Classic did not implement pseudoterminals at the file system layer, so
+`isTTY` cannot be meaningfully implemented.
+
+For most situations like this one, you can guard the affected code with a
+platform conditional and provide a stub implementation:
+
+```diff
+--- a/Sources/Testing/Support/FileHandle.swift
++++ b/Sources/Testing/Support/FileHandle.swift
+
+ var isTTY: Bool {
+ #if SWT_TARGET_OS_APPLE || os(Linux) || os(FreeBSD) || os(Android) || os(WASI)
+   // ...
++#elseif os(Classic)
++  return false
+ #else
+ #warning("Platform-specific implementation missing: cannot tell if a file is a TTY")
+   return false
+ #endif
+ }
+```
+
+If another function in Swift Testing asks if a file is a TTY, it will then
+always get a result of `false` (which is always the correct result on Classic.)
+No further changes are needed in this case.
+
+If your target platform is missing some feature that is used pervasively
+throughout Swift Testing, this approach may be insufficient. Please reach out to
+us in the Swift forums for advice.
+
+## Adding new dependencies
+
+Avoid adding new Swift package or toolchain library dependencies. Swift Testing
+needs to support running tests for all Swift targets except, for the moment, the
+Swift standard library itself. Adding a dependency on another Swift component
+means that that component may be unable to link to Swift Testing. If you find
+yourself needing to link to a Swift component, please reach out to us in the
+Swift forums for advice.
+
+> [!WARNING]
+> Swift Testing has some dependencies on Foundation, specifically to support our
+> JSON event stream. Do not add new uses of Foundation without talking to us
+> first.
```

Comment:
When it comes to Foundation, I would recommend specifically noting that although we do _use_ Foundation in our implementation, we are careful not to _publicly import_ its module, since that would exacerbate existing challenges with that dependency.